pubsub/topic.go

```diff
@@ -585,26 +593,51 @@ func (t *Topic) Publish(ctx context.Context, msg *Message) *PublishResult {
 		Attributes:  msg.Attributes,
 		OrderingKey: msg.OrderingKey,
 	})
+	span.SetAttributes(semconv.MessagingMessagePayloadSizeBytesKey.Int(msgSize))
 
 	t.initBundler()
 	t.mu.RLock()
 	defer t.mu.RUnlock()
 	// TODO(aboulhosn) [from bcmills] consider changing the semantics of bundler to perform this logic so we don't have to do it here
 	if t.stopped {
 		ipubsub.SetPublishResult(r, "", ErrTopicStopped)
+		spanRecordError(span, ErrTopicStopped)
 		return r
 	}
 
+	ctx2, fcSpan := tracer().Start(ctx, publishFlowControlSpanName)
 	if err := t.flowController.acquire(ctx, msgSize); err != nil {
 		t.scheduler.Pause(msg.OrderingKey)
 		ipubsub.SetPublishResult(r, "", err)
+		spanRecordError(span, err)
 		return r
 	}
-	err = t.scheduler.Add(msg.OrderingKey, &bundledMessage{msg, r, msgSize}, msgSize)
-	if err != nil {
+	fcSpan.End()
+
+	_, batchSpan := tracer().Start(ctx2, publishSchedulerSpanName)
+
+	bmsg := &bundledMessage{
+		msg:       msg,
+		res:       r,
+		size:      msgSize,
+		span:      span,
+		batchSpan: batchSpan,
+	}
+
+	if span.SpanContext().IsValid() {
+		if msg.Attributes == nil {
+			msg.Attributes = make(map[string]string)
+		}
+		// Inject the context from the first publish span rather than from flow control / batching.
+		otel.GetTextMapPropagator().Inject(ctx, newMessageCarrier(msg))
```

Comment:
Likewise here, there's a third context being created for tracer().Start(), but neither of them are being used in the text map propagator call.

Reply:
The `Inject` method doesn't take in the later `ctx` since we want the parent span for the subscribe side to be the publish span (the first publish span created)